state/pruner.go

```diff
@@ -0,0 +1,267 @@
+package state
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cometbft/cometbft/config"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/libs/service"
+)
+
+var (
+	AppRetainHeightKey            = []byte("AppRetainHeightKey")
+	CompanionBlockRetainHeightKey = []byte("DCBlockRetainHeightKey")
+	ABCIResultsRetainHeightKey    = []byte("ABCIResRetainHeightKey")
+)
+
+// Pruner is a service that reads the retain heights for blocks, state and ABCI
+// results from the database and prunes the corresponding data based on the
+// minimum retain height set. The service sleeps between each run based on the
+// configured pruner interval, and re-evaluates the retain height.
+type Pruner struct {
+	service.BaseService
+	logger log.Logger
+	// DB to which we save the retain heights
+	bs BlockStore
+	// State store to prune state from
+	stateStore Store
+	interval   time.Duration
+}
+
+type prunerConfig struct {
+	interval time.Duration
+}
+
+func defaultPrunerConfig() *prunerConfig {
+	return &prunerConfig{
+		interval: config.DefaultPruningInterval,
+	}
+}
+
+type PrunerOption func(*prunerConfig)
+
+// PrunerInterval allows control over the interval between each run of the
+// pruner.
+func PrunerInterval(t time.Duration) PrunerOption {
+	return func(p *prunerConfig) { p.interval = t }
+}
+
+func NewPruner(stateStore Store, bs BlockStore, logger log.Logger, options ...PrunerOption) *Pruner {
+	cfg := defaultPrunerConfig()
+	for _, opt := range options {
+		opt(cfg)
+	}
+	p := &Pruner{
+		bs:         bs,
+		stateStore: stateStore,
+		logger:     logger,
+		interval:   cfg.interval,
+	}
+	p.BaseService = *service.NewBaseService(logger, "Pruner", p)
+	return p
+}
+
+func (p *Pruner) OnStart() error {
+	go p.pruningRoutine()
+	return nil
+}
+
+// SetApplicationRetainHeight sets the application retain height with some
+// basic checks on the requested height.
+//
+// If a higher retain height is already set, we cannot accept the requested
+// height because the blocks might have been pruned.
+//
+// If the data companion has already set a retain height to a higher value we
+// also cannot accept the requested height as the blocks might have been
+// pruned.
+func (p *Pruner) SetApplicationRetainHeight(height int64) error {
```

Comment:
This might be unecessary, but both `SetApplicationRetainHeight` and `SetCompanionRetainHeight` are near identical...could consolidate into a generalized `SetRetainHeight` and pass in the relevant function calls as args

```go
func (p *Pruner) SetRetainHeight(
	height int64,
	getServiceRetainHeight func() (int64, error),
	getAlternateRetainHeight func() (int64, error),
	saveServiceRetainHeight func(int642 int64) error,
) error {
	if height <= 0 || height < p.bs.Base() || height > p.bs.Height() {
		return ErrInvalidHeightValue
	}
	currentServiceRetainHeight, err := getServiceRetainHeight()
	if err != nil {
		if err == ErrKeyNotFound {
			currentServiceRetainHeight = height
		} else {
			return err
		}
	}
	currentAltRetainHeight, err := getAlternateRetainHeight()
	var noAltRetainHeight bool
	if err != nil {
		if err == ErrKeyNotFound {
			noAltRetainHeight = true
		} else {
			return err
		}
	}
	if currentServiceRetainHeight > height || (!noAltRetainHeight && currentAltRetainHeight > height) {
		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
	}
	err = saveServiceRetainHeight(height)
	return err
}
``` 

Then call in `execution.go`
```go
err := blockExec.pruner.SetRetainHeight(
			retainHeight,
			blockExec.pruner.stateStore.GetApplicationRetainHeight,
			blockExec.pruner.stateStore.GetCompanionBlockRetainHeight,
			blockExec.pruner.stateStore.SaveApplicationRetainHeight,
		)
```

Reply:
I would argue it probably makes more sense for it to stay the way it is. Duplication often results in better readability and greater simplicity.